pkg/planner/core/indexmerge_unfinished_path.go

```diff
@@ -0,0 +1,416 @@
+// If we failed to collect access filters for one candidate access path, the corresponding element in the return slice
+// will be nil.
+// If we failed to collect access filters for all candidate access paths, this function will return nil.
+/*
+Example1 (consistent with the one in generateUnfinishedIndexMergePathFromORList()):
+  Input:
+    expr: 1 member of j->'$.a'
+    candidateAccessPaths: [idx1(a, j->'$.a' unsigned array), idx2(j->'$.b' unsigned array, a)]
+  Output:
+    [unfinishedAccessPath{idx1,1 member of j->'$.a'}, nil]
+
+Example2:
+  Input:
+    expr: a = 3
+    candidateAccessPaths: [idx1(a, j->'$.a' unsigned array), idx2(j->'$.b' unsigned array, a)]
+  Output:
+    [unfinishedAccessPath{idx1,a=3}, unfinishedAccessPath{idx2,a=3}]
+*/
+func initUnfinishedPathsFromExpr(
+	ds *DataSource,
+	candidateAccessPaths []*util.AccessPath,
+	expr expression.Expression,
+) unfinishedAccessPathList {
+	retValues := make([]unfinishedAccessPath, len(candidateAccessPaths))
+	ret := make([]*unfinishedAccessPath, 0, len(candidateAccessPaths))
+	for i := range candidateAccessPaths {
+		ret = append(ret, &retValues[i])
+	}
+	for i, path := range candidateAccessPaths {
+		ret[i].index = path.Index
+		// case 1: try to use the previous logic to handle non-mv index
+		if !isMVIndexPath(path) {
+			// generateNormalIndexPartialPaths4DNF is introduced for handle a slice of DNF items and a slice of
+			// candidate AccessPaths before, now we reuse it to handle single filter and single candidate AccessPath,
+			// so we need to wrap them in a slice here.
+			paths, needSelection, usedMap := ds.generateNormalIndexPartialPaths4DNF(
+				[]expression.Expression{expr},
+				[]*util.AccessPath{path},
+			)
+			if len(usedMap) == 1 && usedMap[0] && len(paths) == 1 {
+				ret[i].initedAsFinished = true
+				ret[i].accessFilters = paths[0].AccessConds
+				ret[i].needKeepFilter = needSelection
+				continue
+			}
+		}
+		if path.IsTablePath() {
+			continue
+		}
+		idxCols, ok := PrepareIdxColsAndUnwrapArrayType(ds.table.Meta(), path.Index, ds.TblCols, false)
+		if !ok {
+			continue
+		}
+		cnfItems := expression.SplitCNFItems(expr)
+
+		// case 2: try to use the previous logic to handle mv index
+		if isMVIndexPath(path) {
+			accessFilters, remainingFilters, tp := collectFilters4MVIndex(ds.SCtx(), cnfItems, idxCols)
+			if len(accessFilters) > 0 && (tp == multiValuesOROnMVColTp || tp == singleValueOnMVColTp) {
+				ret[i].initedAsFinished = true
+				ret[i].accessFilters = accessFilters
+				ret[i].needKeepFilter = len(remainingFilters) > 0
+				continue
+			}
+		}
+
+		// case 3: use the new logic if the previous logic didn't succeed to collect access filters that can build a
+		// valid range directly.
+		ret[i].idxColHasAccessFilter = make([]bool, len(idxCols))
+		for j, col := range idxCols {
+			for _, cnfItem := range cnfItems {
+				if ok, tp := checkFilter4MVIndexColumn(ds.SCtx(), cnfItem, col); ok &&
```

Comment:
should we change the name of `checkFilter4MVIndexColumn`, since it also handles normal indexes